### DISCUSSION TITLE
How do I enable OSC52 on Mosh and SSH?

### DISCUSSION THREAD

**Discussion by iflowerml** [blinksh/blink#1948] [Q&A] [answered]
Can I copy/paste between a remote machine and my device? How do I enable OSC52 between a remote machine and my device? How can I copy text to the system clipboard across Mosh, SSH, tmux, screen, Emacs, and Vim?

**Comment by iflowerml** [comment, marked as answer]
CLI applications, specially editors, regularly use their clipboard to perform copy or yank operations. [OSC52 is an ANSI control sequence]([https://www.reddit.com/r/vim/comments/k1ydpn/a_guide_on_how_to_copy_text_from_anywhere/](https://www.reddit.com/r/vim/comments/k1ydpn/a_guide_on_how_to_copy_text_from_anywhere/?utm_source=share&utm_medium=web2x&context=3)) used to instruct the terminal to copy the string it carries to the system clipboard, even through remote connections over SSH and Mosh.

Mosh supports the OSC52 sequence out of the box since version 1.4.0, so make sure your mosh-server is up to date. 

Then, you just need to instruct the specific CLI application to make use of the OSC52 sequence. Here is how to do so, and please chime in on this thread to add others!

**Vim:**
To enable copy/paste you’ll need to install[ ojroques/vim-oscyank](https://github.com/ojroques/vim-oscyank) and add the following line to `.vimrc`:
```
autocmd TextYankPost * if v:event.operator is 'y' && v:event.regname is '' | OSCYankReg " | endif"
```
then enable this setting:
```
set clipboard& clipboard^=unnamed,unnamedplus
```
**Emacs:**
To enable copy/paste on Emacs, you can install[ clipetty](https://github.com/spudlyo/clipetty). Check out the GitHub repo for the complete documentation. Enable [use-package](https://github.com/jwiegley/use-package), by adding this to your config:
```
(use-package clipetty
:ensure t
:hook (after-init . global-clipetty-mode))
```
**TMUX:**
To enable copy/paste between tmux and your device, insert the following lines into your `tmux.conf`:
```
set -g set-clipboard on
# allow other apps to passthrough sequences (including OSC52)
set -g allow-passthrough on
set -ag terminal-overrides "vte*:XT:Ms=\\E]52;c;%p2%s\\7,xterm*:XT:Ms=\\E]52;c;%p2%s\\7"
```

**Reply by carloscabanero** [reply]
Other tips to figure out in case you have issues with OSC-52:
- It is important to figure out if the problem is on the terminal or in the application. To do so, on a clean terminal shell, run:
```
printf "\033]52;c;Y29weXBhc3RhIQ==\a\n"
```
And `copypasta!` should show up in the clipboard.
- This is another good link for resources: https://seankhliao.com/blog/12020-05-14-osc-52/

**Comment by shlevy** [comment]
In addition to @iflowerml‘s [answer](https://github.com/blinksh/blink/discussions/1948#discussioncomment-8338646), you can also use the [osc](https://github.com/theimpostor/osc) tool on the command line:

> $ echo "copied by osc" | (unset SSH_TTY; osc copy)

Note that the `unset SSH_TTY` is needed because `mosh-server` is erroneously inheriting this variable from the SSH session; this will likely be fixed in a future version of `blink`

**Reply by carloscabanero** [reply]
Now tracked on #1957

**Comment by kkebo** [comment]
This instruction is very helpful for me and it worked well.

https://sunaku.github.io/tmux-yank-osc52.html

**Comment by xero** [comment]
for anyone looking for examples of osc52 workflows in Linux, my dotfiles are here for your convenience. The readme in my repository [talks more about my unified vps & blink setup](https://github.com/xero/dotfiles?tab=readme-ov-file#vps--local-clipboard)

for neovim, i use the osc52-yank plugin. this updates your yank actions to add the proper escaping.
https://github.com/xero/dotfiles/blob/main/neovim/.config/nvim/lua/plugins/osc52-yank.lua

here's the relevant section of my tmux configuration 
https://github.com/xero/dotfiles/blob/main/tmux/.config/tmux/tmux.conf#L180

my yank script is a wrapper for xsel, pbcopy, tmuxattach, etc, that adds osc52 escape sequences.
https://github.com/xero/dotfiles/blob/main/bin/.local/bin/yank
